src/host_abi.zig

```diff
@@ -0,0 +1,106 @@
+/// All Roc functions that are exposed to the host take 1 argument and return void.
+/// The 1 argument is a pointer to one of these structs, which includes an address
+/// that the Roc call will write the return value into. This design makes Roc's ABI
+/// very simple; the calling convention is just "pass a single pointer".
+pub const RocCall = struct {
+    /// Function pointers that the Roc program uses, e.g. alloc, dealloc, etc.
+    ops: *RocOps,
+    /// The argument that the Roc entrypoint function will receive from the host.
+    /// (If multiple arguments are needed, this should be a pointer to a struct.)
+    arg: *anyopaque,
+    /// What the Roc entrypoint function will return to the host.
+    /// (The caller should have allocated enough space for the Roc function to write
+    /// the entire return value into this.)
+    ret: *anyopaque,
+};
```

Comment:
I wouldn't mind expanding this to reduce overhead a bit, but this is probably fine.

My personal preferences would be that a roc function takes N pointers.
1. Pointer to RocOps
2. Pointer to Ret (null if void)
3. N Pointers, 1 for each arg

This would be just as easy to handle from the ABI perspective and it avoids some of the extra repacking of pointers that is required for this to work.

Ultimately probably not that big of a deal.

Reply:
Probably the biggest concern with the suggestion above is that arg is a struct. That is likely to actually be problematic in practice.

Imagine I have the function `LargeStruct, Enum -> LargeStruct` (maybe an elm architecture update function on a model that is exposed to the host).

The host abi with args needing to be a single struct would be `{LargeStruct, Enum} -> LargeStruct`. Creating `{LargeStruct, Enum}` will incur a `memcpy` on `LargeStruct` for every call to this function. So I think we want an extra layer of indirection here to avoid the copy.

To remove that copy, we could instead do any of the following
1. Inline the N arg pointers like I mentioned above
2. Change from a struct of data to a struct of pointers
3. Use some sort of list of pointers or other indirection

Reply:
Makes sense! I changed this in ea748bb2286bab3739a4644f84aedeb535aba623 - what do you think?

Reply:
Yeah, that makes sense. And in zig the `anytype` would map to a struct of pointers essentially for all of the arg pointers. So like if you have 2 i64 inputs, you would make the `anytype` be `{*I64, *I64}`. That sounds like it works. And for the example above it would be `{*LargeStruct, *Enum}`